### PR TITLE
Handle unusual maxsigcachesize gracefully

### PR DESCRIPTION
**Pull request by jnewbery**
belt-and-braces approach:

- first commit fixes the bug where maxsigcachesize being zero causes segfault
- second commit handles maxsigcachesize being negative by setting it to zero if a negative number is provided
- third commit handles maxsigcachesize being too large by setting it to MAX_MAX_SIG_CACHE_SIZE if an unreasonably large number is provided.

I've set `MAX_MAX_SIG_CACHE_SIZE` absurdly high (16GB) for now, so setting maxsigcachesize to anything unreasonably large will still cause a bad_alloc for most users. Input welcomed on what a more reasonable `MAX_MAX_SIG_CACHE_SIZE` should be, or whether we should leave it out entirely.

Replaces https://github.com/bitcoin/bitcoin/pull/9770

### CODE REVIEW

**Comment by laanwj** [comment]
> I've set MAX_MAX_SIG_CACHE_SIZE absurdly high (16GB) for now, so setting maxsigcachesize to anything unreasonably large will still cause a bad_alloc for most users. Input welcomed on what a more reasonable MAX_MAX_SIG_CACHE_SIZE should be, or whether we should leave it out entirely.

I'd say that is reasonable. We don't want bad_alloc for negative amounts (which you fixed), but if the user specifies a higher value than available memory then it's what you expect. Even better error handling would catch the bad_alloc and print a friendly error message, but that's overdoing it in this particular case.

**Comment by JeremyRubin** [comment]
utACK e6374a0.

I would say it'd also be fine to restrict to ~512MB (covers at least a full day of maximum sigop blocks) if you want to be more aggressive about it, but 16GB is fine too.

**Comment by sipa** [comment]
utACK

**Comment by morcos** [comment]
utACK e6374a0

**Comment by jtimon** [comment]
utACK e6374a0, perhaps some squashing can be done, at least the first 2 commits. Not a big deal though.

File: src/script/sigcache.cpp [hunk 1]
```diff
@@ -95,7 +95,7 @@ void InitSignatureCache()
 {
     // nMaxCacheSize is unsigned. If -maxsigcachesize is set to zero,
     // setup_bytes creates the minimum possible cache (2 elements).
-    size_t nMaxCacheSize = std::max((int64_t)0, GetArg("-maxsigcachesize", DEFAULT_MAX_SIG_CACHE_SIZE)) * ((size_t) 1 << 20);
+    size_t nMaxCacheSize = std::min(std::max((int64_t)0, GetArg("-maxsigcachesize", DEFAULT_MAX_SIG_CACHE_SIZE)), MAX_MAX_SIG_CACHE_SIZE) * ((size_t) 1 << 20);
```
Comment (jtimon) [inline src/script/sigcache.cpp]: Maybe ```((size_t) 1 << 20)``` could be made a constant/macro, not necessarily in this PR.

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/9777/commits/e6374a03b9869362595acd5cc5e5b8994b244af1

`-maxsigcachesize=0` does not result in a crash anymore. Max limit applies (=20000 results in 16384).

+1 for squashing.

**Review by gmaxwell — APPROVED** [review]
utACK  16384 is implausibly large but fine.  Reducing memory option footguns is something for a future time when we've solved the nearly infinite number of more pressing issues, and is better than doing nothing there.

I believe the most common case of a large value though is misunderstanding the units (e.g. giving it a limit in bytes).

**Comment by jnewbery** [comment]
> I believe the most common case of a large value though is misunderstanding the units (e.g. giving it a limit in bytes).

Or using a pre-2016 bitcoin.conf when maxsigcachesize was measured in cache entries instead of MiB :(

**Comment by gmaxwell** [comment]
ouch. uh. that is an argument for throwing an error or renaming the argument rather than just claiming to an insane value.

**Comment by laanwj** [comment]
> ouch. uh. that is an argument for throwing an error or renaming the argument rather than just claiming to an insane value.

Yes it also would have been perfectly valid to make `InitSignatureCache()` return an error status, and handle the error upstream by printing a message (hopefully a generalized "value too large" one instead of adding a specific translation message) instead of clamping it here.

However now that we have this, I don't see a point in redesigning it again and going through another review. As you say there are an infinite number of more pressing issues.

**Comment by laanwj** [comment]
squashed d72fe44fc1989575ff5de98c882c0010d1de26e7..e6374a03b9869362595acd5cc5e5b8994b244af1 into 55c403b8febe02555c52bac7028cd6b1f006fad1, merged via 8dee8221770893fbf0ec6c19ad385537984ba44f

**Pull request closed without merging**